### DISCUSSION TITLE
Why poolclass needs to use NullPool

### DISCUSSION THREAD

**Discussion by pcorvoh** [sqlalchemy/sqlalchemy#9639] [Usage Questions]
Description：
1、I'm using fastapi to execute sqlalchemy, after building a program and requesting the port, sqlalchemy prompts `got Future <Future pending> attached to a different loop`, I was checking the documentation and found that this may be related to multiple events in the thread pool, I'd like to know why asynchronous would have this problem
```python
async_engine = create_async_engine(
        BaseConfig.ASYNC_SQLALCHEMY_URI,
        pool_recycle=1500,
        poolclass=NullPool,
        # Enable pool_pre_ping for connection health checks
        pool_pre_ping=True,
)
```
It may have something to do with this [sqlalchemy docs](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html)

But when I use my own defined thread pool, I need to pass a loop event so that my whole fastapi project stays in the same loop, but I don't know how to do it

coding:
```python
import asyncio
from asyncio import Future
from sqlalchemy.pool import _ConnectionFairy, QueuePool


class CustomAsyncQueuePool(QueuePool):

    def __init__(self, *args, loop=None, **kwargs):
        super().__init__(*args, **kwargs)
        self.loop = loop
        print(f"CustomAsyncQueuePool loop: {self.loop}")

    async def acquire(self):
        while True:
            try:
                waiter = asyncio.Future(
                        loop=self.loop
                )  # Use asyncio.Future().
                self._wq.put(waiter)
                with self._lock:
                    if self._max_overflow > -1 and len(self._overflow
                                                       ) >= self._max_overflow:
                        if self._timeout is None:
                            return await waiter
                        else:
                            timeout_fut = asyncio.ensure_future(
                                    asyncio.sleep(self._timeout),
                                    loop=self.loop  # Pass the same event loop.
                            )
                            try:
                                await asyncio.wait([waiter, timeout_fut],
                                                   timeout=self._timeout)
                            except asyncio.TimeoutError:
                                with self._lock:
                                    self._overflow.remove(waiter)
                                self.logger.warning(
                                        "QueuePool limit of size %d overflow %d reached, connection timed out, timeout=%d",
                                        self.size(), len(self._overflow),
                                        self._timeout
                                )
                                return None
                    else:
                        break
            except Exception as e:
                del waiter
                raise

        try:
            # Get a free connection.
            with self._dec_overflow_pending():
                con = self._pool.get()
            try:
                # Create a connection fairy object.
                cf = _ConnectionFairy(self, con)
                self._inc_overflow()
                fut = asyncio.Future(loop=self.loop)  # Use asyncio.Future().
                result = cf._async_get_conn(fut)  # Pass the Future object.
                await fut  # Wait for the Future to complete.
                return result
            except:
                with self._dec_overflow_pending():
                    self._do_return_conn(con)

        except Exception as e:
            self.logger.error(
                    "Connection exception occurred when acquiring pooled connection."
            )
  
async_engine = create_async_engine(
        BaseConfig.ASYNC_SQLALCHEMY_URI,
        pool_recycle=1500,
        poolclass=CustomAsyncQueuePool,
        # Enable pool_pre_ping for connection health checks
        pool_pre_ping=True,
)
```
I've checked the documentation and can't find what I'm looking for, I want a perfect solution because I need thread pooling, can you explain to me why the documentation only eliminates thread pooling and there is no alternative (of course I didn't look at the underlying design of sqlalchemy)

**Comment by CaselIT** [comment]
Hi,

It's not clear to me what the problem is and what are you trying to solve using `CustomAsyncQueuePool`?

**Reply by pcorvoh** [reply]
I defined a link pool for CustomAsyncQueuePool and I defined an additional parameter loop, but how do I pass this loop to the CustomAsyncQueuePool class via async_create_engine and why do I have to use the poolclass= NullPool class so that I don't get the error `got Future <Future pending> attached to a different loop`, otherwise I will get an error if I use asynchronous to operate the database, multiple loop events

**Reply by CaselIT** [reply]
You will need to pass the pool instance to create_async_engine, instead of `poolclass` use `pool`.

> why do I have to use the poolclass= NullPool class so that I don't get the error got Future <Future pending> attached to a different loop, otherwise I will get an error if I use asynchronous to operate the database, multiple loop events

I'm not aware of any issue that would cause that error. Please provide an example that raises that error

**Reply by pcorvoh** [reply]
> >您需要将池实例传递给create_async_engine，而不是使用 .`poolclass``pool`

I don't seem to understand, can you give a minimal case, thank you very much

**Reply by CaselIT** [reply]
> > > 您需要将池实例传递给create_async_engine，而不是使用 .`` poolclass``pool ``
> 
> I don't seem to understand, can you give a minimal case, thank you very much

You can pass an instance of the pool doing
```py
async_engine = create_async_engine(
        BaseConfig.ASYNC_SQLALCHEMY_URI,
        # Enable pool_pre_ping for connection health checks
        pool=CustomAsyncQueuePool(loop=..., re,recycle=1500, pre_ping=True)
)
```

**Reply by pcorvoh** [reply]
> > > > 您需要将池实例传递给create_async_engine，而不是使用 .` poolclass`pool ``
> > 
> > 
> > I don't seem to understand, can you give a minimal case, thank you very much
> 
> You can pass an instance of the pool doing
> 
> ```python
> async_engine = create_async_engine(
>         BaseConfig.ASYNC_SQLALCHEMY_URI,
>         # Enable pool_pre_ping for connection health checks
>         pool=CustomAsyncQueuePool(loop=..., re,recycle=1500, pre_ping=True)
> )
> ```

Thank you very much, I think I already know what the problem is, thank you for your patience!
